src/Xrefcheck/Verify.hs

```diff
@@ -231,22 +275,50 @@ data NeedsCaching key
 -- then the @action@ will not be executed, and the value is added to the accumulator list.
 -- After the whole list has been traversed, the accumulator is traversed once again to ensure
 -- every asynchronous action is completed.
+-- If interrupted by AsyncException, returns this exception and list of already calcualted results
+-- (their subset can be arbitrary). Computations that were not ended till this moment are cancelled.
 forConcurrentlyCaching
-  :: Ord cacheKey
-  => [a] -> (a -> NeedsCaching cacheKey) -> (a -> IO b) -> IO [b]
+  :: forall a b cacheKey.  Ord cacheKey
+  => [a] -> (a -> NeedsCaching cacheKey) -> (a -> IO b) -> IO (Either (AsyncException, [b]) [b])
 forConcurrentlyCaching list needsCaching action = go [] M.empty list
   where
-    go acc cached (x : xs) = case needsCaching x of
-      NoCaching -> do
-        withAsync (action x) $ \b ->
-          go (b : acc) cached xs
-      CacheUnderKey cacheKey -> do
-        case M.lookup cacheKey cached of
-          Nothing -> do
+    go
+      :: [Async b]
+      -> Map cacheKey (Async b)
+      -> [a]
+      -> IO (Either (AsyncException, [b]) [b])
+    go acc cached items =
+      case items of
+
+        (x : xs) -> case needsCaching x of
+          NoCaching -> do
             withAsync (action x) $ \b ->
-              go (b : acc) (M.insert cacheKey b cached) xs
-          Just b -> go (b : acc) cached xs
-    go acc _ [] = for acc wait <&> reverse
+              go (b : acc) cached xs
+          CacheUnderKey cacheKey -> do
+            case M.lookup cacheKey cached of
+              Nothing -> do
+                withAsync (action x) $ \b ->
+                  go (b : acc) (M.insert cacheKey b cached) xs
+              Just b -> go (b : acc) cached xs
+
+        [] -> handleJust
+                (\case
```

Comment:
Excessive indentation here too, it should be:

```hs
[] ->
  handleJust
    (\case ...
```

Or

```hs
[] -> handleJust
  (\case ...
```

Reply:
both looks a bit strange for me, but ok